Conversion panic: `each.value seen during expression conversion, but each scope not set`

Converting https://github.com/aztfmod/terraform-azurerm-caf

```
pulumi convert --from terraform --language typescript --out pulumi
Converting from terraform...
warning: Plugin converters are currently experimental
warning: failed to install provider "external": could not find latest version for provider external: 404 HTTP error fetching plugin from https://api.github.com/repos/pulumi/pulumi-external/releases/latest. If this is a private GitHub repository, try providing a token via the GITHUB_TOKEN environment variable. See: https://github.com/settings/tokens
warning: failed to install provider "azurecaf": could not find latest version for provider azurecaf: 404 HTTP error fetching plugin from https://api.github.com/repos/pulumi/pulumi-azurecaf/releases/latest. If this is a private GitHub repository, try providing a token via the GITHUB_TOKEN environment variable. See: https://github.com/settings/tokens
warning: failed to install provider "azapi": could not find latest version for provider azapi: 404 HTTP error fetching plugin from https://api.github.com/repos/pulumi/pulumi-azapi/releases/latest. If this is a private GitHub repository, try providing a token via the GITHUB_TOKEN environment variable. See: https://github.com/settings/tokens
warning: failed to install provider "null": could not find latest version for provider null: 404 HTTP error fetching plugin from https://api.github.com/repos/pulumi/pulumi-null/releases/latest. If this is a private GitHub repository, try providing a token via the GITHUB_TOKEN environment variable. See: https://github.com/settings/tokens
panic: fatal: A failure has occurred: each.value seen during expression conversion, but each scope not set

goroutine 49 [running]:

github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.69.1-0.20230605220215-9df9a4a8bd14/go/common/util/contract/failfast.go:23

github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf({0x1060c8923?, 0x1400314ec90?}, {0x0?, 0x4?, 0x1068398a0?})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.69.1-0.20230605220215-9df9a4a8bd14/go/common/util/contract/fail.go:32 +0xbc

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.rewriteTraversal(0xc0?, 0x14001bba410, {0x0?, 0x0?}, {0x140032e46c0?, 0x3, 0x4})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1074 +0x1f4c

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertScopeTraversalExpr(...)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1127

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertExpression(0x140010599e0?, 0x5?, {0x0?, 0x10602d0e4?}, {0x106afa248?, 0x14001151aa0?})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1351 +0x128

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertObjectConsExpr(0x1067f6560?, 0x7000140012b3068?, {0x0?, 0x0?}, 0x14001fb6140?)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:850 +0x2c4

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertExpression(0x14001bb92d8?, 0x10604e5e7?, {0x0?, 0x14000620ba0?}, {0x106afa168?, 0x14001fb61e0?})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1341 +0x68

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertProvisioner(0x107b2ba30?, {0x0?, 0x0?}, 0x2?, 0x1400214ac00, {0x140011be820, 0xf}, 0x0, 0x14001963578)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1903 +0x998

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.convertManagedResources(0x1067837e0?, {0x106ae0e80, 0x140002a6700}, 0x14001bba410, 0x140006e8300, 0x14001963578)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:1982 +0xc84

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.translateModuleSourceCode(0x10677e9e0?, {0x106b051e0, 0x107b2ba30}, {0x14000c80500, 0x9e}, {0x106b05160?, 0x14000a703a0?}, {0x140026284c0, 0x33}, {0x106ae0e80?, ...})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:2870 +0x27dc

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.translateModuleSourceCode(0x14000298070?, {0x106b051e0, 0x107b2ba30}, {0x14000538150, 0x6b}, {0x106b05160?, 0x14000a703a0?}, {0x10652bad8, 0x1}, {0x106ae0e80?, ...})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:2656 +0x428c

github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tf2pulumi/convert.TranslateModule(...)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.48.1-0.20230606121021-a8160de9d11a/pkg/tf2pulumi/convert/tf.go:2926

main.(*tfConverter).ConvertProgram(0x14000b41718, {0x104b23f28?, 0x14000b41748?}, 0x14000bbe540)

    /Users/justin/go/src/github.com/pulumi/pulumi-converter-terraform/cmd/pulumi-converter-terraform/main.go:57 +0x1c8

github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.(*converterServer).ConvertProgram(0x140005b5320, {0x106af96b0, 0x14000bbe510}, 0x14000bc6120)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.69.1-0.20230605220215-9df9a4a8bd14/go/common/resource/plugin/converter_server.go:64 +0xec

github.com/pulumi/pulumi/sdk/v3/proto/go._Converter_ConvertProgram_Handler.func1({0x106af96b0, 0x14000bbe510}, {0x10697a4e0?, 0x14000bc6120})

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.69.1-0.20230605220215-9df9a4a8bd14/proto/go/converter_grpc.pb.go:122 +0x74

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x106af96b0, 0x14000bbe1b0}, {0x10697a4e0, 0x14000bc6120}, 0x14000a700a0, 0x140004ce1c8)

    /Users/justin/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x308

github.com/pulumi/pulumi/sdk/v3/proto/go._Converter_ConvertProgram_Handler({0x106823bc0?, 0x140005b5320}, {0x106af96b0, 0x14000bbe1b0}, 0x140009ae070, 0x140005b5180)

    /Users/justin/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.69.1-0.20230605220215-9df9a4a8bd14/proto/go/converter_grpc.pb.go:124 +0x138

google.golang.org/grpc.(*Server).processUnaryRPC(0x14000340960, {0x106b030c0, 0x140006036c0}, 0x14000332fc0, 0x14000aef9b0, 0x107ab4098, 0x0)

    /Users/justin/go/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:1345 +0xc90

google.golang.org/grpc.(*Server).handleStream(0x14000340960, {0x106b030c0, 0x140006036c0}, 0x14000332fc0, 0x0)

    /Users/justin/go/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:1722 +0x82c

google.golang.org/grpc.(*Server).serveStreams.func1.2()

    /Users/justin/go/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:966 +0x84

created by google.golang.org/grpc.(*Server).serveStreams.func1

    /Users/justin/go/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:964 +0x28c

error: rpc error: code = Unavailable desc = error reading from server: EOF
```

--- a/pkg/tf2pulumi/convert/testdata/provisioners_for_each/experimental_pcl/main.pp
+++ b/pkg/tf2pulumi/convert/testdata/provisioners_for_each/experimental_pcl/main.pp
@@ -1,0 +1,24 @@
+# Test for provisioners feature in Terraform https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax
+# In combination with for_each
+
+
+config "echoData" "map(string)" {
+  default = {
+    first  = "First"
+    second = "Second"
+  }
+}
+resource "localExecResource" "simple:index:resource" {
+  options {
+    range = echoData
+  }
+  inputOne = "hello"
+  inputTwo = true
+}
+resource "localExecResourceProvisioner0" "command:local:Command" {
+  options {
+    range     = echoData
+    dependsOn = localExecResource
+  }
+  create = "echo ${range.value}"
+}

--- a/pkg/tf2pulumi/convert/testdata/provisioners_for_each/main.tf
+++ b/pkg/tf2pulumi/convert/testdata/provisioners_for_each/main.tf
@@ -1,0 +1,23 @@
+# Test for provisioners feature in Terraform https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax
+# In combination with for_each
+
+#if EXPERIMENTAL
+
+variable "echo_data" {
+    type = map(string)
+    default = {
+        "first": "First"
+        "second": "Second"
+    }
+}
+
+resource "simple_resource" "local_exec_resource" {
+    for_each = var.echo_data
+    input_one = "hello"
+    input_two = true
+
+    provisioner "local-exec" {
+        command = "echo ${each.value}"
+    }
+}
+#endif

--- a/pkg/tf2pulumi/convert/tf.go
+++ b/pkg/tf2pulumi/convert/tf.go
@@ -1880,6 +1880,7 @@ func convertProvisioner(
 	info il.ProviderInfoSource, scopes *scopes,
 	provisioner *configs.Provisioner,
 	resourceName string, provisionerIndex int,
+	forEach hcl.Expression,
 	target *hclwrite.Body,
 ) {
 	if provisioner.Type != "local-exec" {
@@ -1902,16 +1903,31 @@ func convertProvisioner(
 	optionsBlock := blockBody.AppendNewBlock("options", nil)
 	optionsBlockBody := optionsBlock.Body()
 
+	if forEach != nil {
+		forEachExpr := convertExpression(sources, scopes, "", forEach)
+		scopes.eachKey = hcl.Traversal{hcl.TraverseRoot{Name: "range"}, hcl.TraverseAttr{Name: "key"}}
+		scopes.eachValue = hcl.Traversal{hcl.TraverseRoot{Name: "range"}, hcl.TraverseAttr{Name: "value"}}
+		optionsBlockBody.SetAttributeRaw("range", forEachExpr)
+	}
+
 	// The first provisioner dependsOn the resource we're provisioning, each provisioner after that depends on
 	// the previous provisioner
-	dependsOn := hclwrite.Tokens{makeToken(hclsyntax.TokenOBrack, "[")}
+	var dependsOn hclwrite.Tokens
+	if forEach == nil {
+		dependsOn = append(dependsOn, makeToken(hclsyntax.TokenOBrack, "["))
+	}
+
 	if provisionerIndex == 0 {
 		dependsOn = append(dependsOn, makeToken(hclsyntax.TokenIdent, resourceName))
 	} else {
 		dependsOn = append(dependsOn, makeToken(hclsyntax.TokenIdent,
 			fmt.Sprintf("%sProvisioner%d", resourceName, (provisionerIndex-1))))
 	}
-	dependsOn = append(dependsOn, makeToken(hclsyntax.TokenCBrack, "]"))
+
+	if forEach == nil {
+		dependsOn = append(dependsOn, makeToken(hclsyntax.TokenCBrack, "]"))
+	}
+
 	optionsBlockBody.SetAttributeRaw("dependsOn", dependsOn)
 
 	attributes, _ := provisioner.Config.JustAttributes()
@@ -2003,7 +2019,7 @@ func convertManagedResources(sources map[string][]byte,
 
 	// Add "command:Command" resources to handle provisioners
 	for idx, provisioner := range managedResource.Managed.Provisioners {
-		convertProvisioner(sources, info, scopes, provisioner, pulumiName, idx, target)
+		convertProvisioner(sources, info, scopes, provisioner, pulumiName, idx, managedResource.ForEach, target)
 	}
 }
 

